From ae845404fef1424e7715e199e8186e60c744dc5d Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Sat, 9 Aug 2025 15:41:56 +0100
Subject: [PATCH 6/6] Provide correct FIPS 140-3 approved service indicator

bouncycastle approved_only security setting is the indicator of the
approved only mode, which is significant for compliance.

the xpack fips mode setting, actually doesn't do anything to turn on
or off fips, mode as it only controls the required license type and
does rudementary heuristic checks, none of which indicative if
validated implementation is in use.
---
 .../xpack/security/SecurityUsageTransportAction.java         | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/SecurityUsageTransportAction.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/SecurityUsageTransportAction.java
index 15b758ceb49..1a78ce54a97 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/SecurityUsageTransportAction.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/SecurityUsageTransportAction.java
@@ -13,6 +13,7 @@ import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Booleans;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.injection.guice.Inject;
 import org.elasticsearch.license.XPackLicenseState;
@@ -36,6 +37,7 @@ import org.elasticsearch.xpack.security.operator.OperatorPrivileges;
 import org.elasticsearch.xpack.security.profile.ProfileService;
 import org.elasticsearch.xpack.security.transport.filter.IPFilter;
 
+import java.security.Provider;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
@@ -217,7 +219,8 @@ public class SecurityUsageTransportAction extends XPackUsageFeatureTransportActi
         return ipFilter.usageStats();
     }
 
+    // Ensure JVM is in approved only mode, as FIPS_MODE_ENABLED is a freeform setting that doesn't actually indicate approved status
     static Map<String, Object> fips140Usage(Settings settings) {
-        return singletonMap("enabled", FIPS_MODE_ENABLED.get(settings));
+        return singletonMap("enabled", Booleans.parseBoolean(java.security.Security.getProperty("org.bouncycastle.fips.approved_only"), false) );
     }
 }
-- 
2.48.1

